File: src/GlassExtended2.cpp
// this file is covered by the  GNU LESSER GENERAL PUBLIC LICENSE Version 3 or later
// please see LICENSE.txt for more details and licensing issues
// copyright Etienne de Foras ( the author )  mailto: etienne.deforas@gmail.com

#include "GlassExtended2.h"
#include <cmath>
#include <cassert>

// glass definition as in the https://github.com/nzhagen/zemaxglass

//////////////////////////////////////////////////////////////////////////////
GlassExtended2::GlassExtended2()
{
    _sName="Extended2";
    _sFormula="Extended2";

    _dC0=0.;
    _dC1=0.;
    _dC2=0.;
    _dC3=0.;
    _dC4=0.;
    _dC5=0.;
    _dC6=0.;
    _dC7=0.;
}
//////////////////////////////////////////////////////////////////////////////
Glass* GlassExtended2::clone() const
{
    return new GlassExtended2(*this);
}
//////////////////////////////////////////////////////////////////////////////
GlassExtended2::GlassExtended2(const GlassExtended2& m):
    Glass(m)
{
    assert(m._sFormula=="Extended2");

    _dC0=m._dC0;
    _dC1=m._dC1;
    _dC2=m._dC2;
    _dC3=m._dC3;
    _dC4=m._dC4;
    _dC5=m._dC5;
    _dC6=m._dC6;
    _dC7=m._dC7;
}
//////////////////////////////////////////////////////////////////////////////
double GlassExtended2::compute_index(double dLambdaMicrons)
{
    //use the Extended2 formulae
    //	n^2 = c0 + c1 λ^2 + c2 λ^−2 + c3 λ^−4 + c4 λ^−6 + c5 λ^−8 + c6 λ^4 + c7 λ^6
    
	double l2=dLambdaMicrons*dLambdaMicrons;
	double l4=l2*l2;
	double invl2=1./l2;
	double invl4=invl2*invl2;
	
    //todo optimize with the Horner rule
    double n2=_dC0+_dC1*l2+_dC2*invl2+_dC3*invl4+_dC4*invl2*invl4+_dC5*invl4*invl4+_dC6*l4+_dC7*l4*l2;
    return sqrt(n2);
}
//////////////////////////////////////////////////////////////////////////////
void GlassExtended2::set_coefs(double dC0,double dC1,double dC2,double dC3,double dC4,double dC5,double dC6,double dC7)
{
    _dC0=dC0;
    _dC1=dC1;
    _dC2=dC2;
    _dC3=dC3;
    _dC4=dC4;
    _dC5=dC5;
    _dC6=dC6;
    _dC7=dC7;
}
//////////////////////////////////////////////////////////////////////////////

File: python/GlassExtended2.py
# this file is covered by the  GNU LESSER GENERAL PUBLIC LICENSE Version 3 or later
# please see LICENSE.txt for more details and licensing issues
# copyright Etienne de Foras ( the author )  mailto: etienne.deforas@gmail.com

import math

# glass definition as in the https://github.com/nzhagen/zemaxglass

class GlassExtended2:
    def __init__(self):
        self.name = "Extended2"
        self.formula = "Extended2"

        self.coefficient_0 = 0.0
        self.coefficient_1 = 0.0
        self.coefficient_2 = 0.0
        self.coefficient_3 = 0.0
        self.coefficient_4 = 0.0
        self.coefficient_5 = 0.0
        self.coefficient_6 = 0.0
        self.coefficient_7 = 0.0

    def clone(self):
        return GlassExtended2(self)

    def __init__(self, other):
        assert other.formula == "Extended2"

        self.coefficient_0 = other.coefficient_0
        self.coefficient_1 = other.coefficient_1
        self.coefficient_2 = other.coefficient_2
        self.coefficient_3 = other.coefficient_3
        self.coefficient_4 = other.coefficient_4
        self.coefficient_5 = other.coefficient_5
        self.coefficient_6 = other.coefficient_6
        self.coefficient_7 = other.coefficient_7

    def compute_index(self, lambda_microns):
        # use the Extended2 formulae
        # n^2 = c0 + c1 λ^2 + c2 λ^−2 + c3 λ^−4 + c4 λ^−6 + c5 λ^−8 + c6 λ^4 + c7 λ^6
        
        lambda_squared = lambda_microns * lambda_microns
        lambda_fourth = lambda_squared * lambda_squared
        inverse_lambda_squared = 1.0 / lambda_squared
        inverse_lambda_fourth = inverse_lambda_squared * inverse_lambda_squared
        
        # todo optimize with the Horner rule
        n_squared = (self.coefficient_0 + self.coefficient_1 * lambda_squared +
                      self.coefficient_2 * inverse_lambda_squared +
                      self.coefficient_3 * inverse_lambda_fourth +
                      self.coefficient_4 * inverse_lambda_squared * inverse_lambda_fourth +
                      self.coefficient_5 * inverse_lambda_fourth * inverse_lambda_fourth +
                      self.coefficient_6 * lambda_fourth +
                      self.coefficient_7 * lambda_fourth * lambda_squared)
        return math.sqrt(n_squared)

    def set_coefficients(self, coefficient_0, coefficient_1, coefficient_2, coefficient_3, coefficient_4, coefficient_5, coefficient_6, coefficient_7):
        self.coefficient_0 = coefficient_0
        self.coefficient_1 = coefficient_1
        self.coefficient_2 = coefficient_2
        self.coefficient_3 = coefficient_3
        self.coefficient_4 = coefficient_4
        self.coefficient_5 = coefficient_5
        self.coefficient_6 = coefficient_6
        self.coefficient_7 = coefficient_7

        self.compute_solid_color()